remote-config/src/fetch/fetcher.rs

```diff
@@ -0,0 +1,316 @@
+use std::collections::HashMap;
+use std::sync::atomic::{AtomicU32, Ordering};
+use std::sync::{Arc, Mutex, MutexGuard};
+use base64::Engine;
+use hyper::http::uri::{PathAndQuery, Scheme};
+use hyper::{Body, Client, StatusCode};
+use sha2::{Digest, Sha256, Sha512};
+use tracing::{debug, trace, warn};
+use datadog_trace_protobuf::remoteconfig::{ClientGetConfigsRequest, ClientGetConfigsResponse, ClientState, ClientTracer, ConfigState, TargetFileHash, TargetFileMeta};
+use ddcommon::{connector, Endpoint};
+use crate::{RemoteConfigPath, Target};
+use crate::targets::TargetsList;
+
+const PROD_INTAKE_SUBDOMAIN: &str = "config";
+
+/// Manages files.
+/// Presents store() and update() operations.
+/// It is recommended to minimize the overhead of these operations as they
+pub trait FileStorage {
+    type StoredFile;
+
+    /// A new, currently unknown file was received.
+    fn store(&self, version: u64, path: RemoteConfigPath, contents: Vec<u8>) -> anyhow::Result<Arc<Self::StoredFile>>;
```

Comment:
I wouldn't make a distinction between the 2 calls, the store knows if it's already present or not, what's the different use case?

Reply:
Actually, the ConfigFileStorage impl in shm_remote_config.rs does not maintain it. The implementation is very straightforward there. It's done that way to avoid tracking state and expiration manually in the FileStorage implementations - that task is up to the ConfigFetcherState.